File: python/trieshake/cleaner.py
"""Cleaner module — remove empty/obsolete directories after moves."""

from pathlib import Path


def cleanup_directories(
    base_dir: Path, extension: str | None = None
) -> list[Path]:
    """Remove directories that are empty or contain no files matching extension.

    Directories are removed deepest-first.

    Args:
        base_dir: The root directory.
        extension: If set, remove dirs that have no files with this extension.

    Returns:
        List of removed directory paths (relative to base_dir).
    """
    removed = []

    # Collect all directories, sorted deepest-first
    all_dirs = sorted(
        [d for d in base_dir.rglob("*") if d.is_dir()],
        key=lambda p: len(p.parts),
        reverse=True,
    )

    for d in all_dirs:
        if d == base_dir:
            continue

        # Check if directory has any non-hidden files
        has_matching = False
        for f in d.iterdir():
            if f.name.startswith("."):
                continue
            if f.is_file():
                if extension is None:
                    has_matching = True
                    break
                if f.name.lower().endswith(extension.lower()):
                    has_matching = True
                    break
            if f.is_dir():
                # Still has subdirectories — skip
                has_matching = True
                break

        if not has_matching:
            rel = d.relative_to(base_dir)
            try:
                # Remove hidden files first (e.g. .DS_Store)
                for hidden in d.iterdir():
                    if hidden.is_file() and hidden.name.startswith("."):
                        hidden.unlink()
                d.rmdir()
                removed.append(rel)
            except OSError:
                pass  # Directory not empty or permission error

    return removed

File: trieshake-clj/src/trieshake/cleaner.clj
(ns trieshake.cleaner
  "Cleaner module — remove empty/obsolete directories after moves."
  (:require [clojure.java.io :as io]
            [clojure.string :as str]))

(defn cleanup-directories!
  "Remove directories that are empty or contain no files matching extension.
   Directories are removed deepest-first.
   Returns list of removed directory paths (relative to base-dir)."
  [base-dir extension]
  (let [base (io/file base-dir)
        base-path (.toPath base)
        ;; Collect all directories, sorted deepest-first
        all-dirs (->> (file-seq base)
                      (filter #(.isDirectory %))
                      (remove #(= (.getCanonicalPath %) (.getCanonicalPath base)))
                      (sort-by #(.getNameCount (.toPath %))
                               >))]
    (reduce
     (fn [removed d]
       (let [contents (seq (.listFiles d))
             ;; Check if directory has any non-hidden files or subdirs
             has-matching (some (fn [f]
                                 (cond
                                   (str/starts-with? (.getName f) ".") false
                                   (.isDirectory f) true
                                   (.isFile f) (if extension
                                                 (str/ends-with?
                                                  (str/lower-case (.getName f))
                                                  (str/lower-case extension))
                                                 true)
                                   :else false))
                               contents)]
         (if has-matching
           removed
           (let [rel (str (.relativize base-path (.toPath d)))]
             (try
               ;; Remove hidden files first (e.g. .DS_Store)
               (doseq [f (.listFiles d)
                        :when (and (.isFile f) (str/starts-with? (.getName f) "."))]
                 (.delete f))
               (.delete d)
               (conj removed rel)
               (catch Exception _
                 removed))))))
     []
     all-dirs)))
